File: lib/variant/standard/variant.js
const { Unit, Board } = require('../../board/package')
const Variant = require('../variant')
const { Rule, MilitaryBranch: { Army, Fleet }, State, Phase: { Movement }, ProvinceStatus } =
  require('./rule/package')
const { Power, map, location: $ } = require('./map/package')
const { Turn, Season: { Spring } } = require('./board/package')

const board = new Board(
  map, new State(new Turn(1901, Spring), Movement),
  [
    new Unit(Army, $.Vie, Power.Austria), new Unit(Army, $.Bud, Power.Austria),
    new Unit(Fleet, $.Tri, Power.Austria),
    new Unit(Fleet, $.Edi, Power.England), new Unit(Fleet, $.Lon, Power.England),
    new Unit(Army, $.Lvp, Power.England),
    new Unit(Fleet, $.Bre, Power.France), new Unit(Army, $.Mar, Power.France),
    new Unit(Army, $.Par, Power.France),
    new Unit(Fleet, $.Kie, Power.Germany), new Unit(Army, $.Ber, Power.Germany),
    new Unit(Army, $.Mun, Power.Germany),
    new Unit(Army, $.Ven, Power.Italy), new Unit(Army, $.Rom, Power.Italy),
    new Unit(Fleet, $.Nap, Power.Italy),
    new Unit(Fleet, $.Sev, Power.Russia), new Unit(Army, $.Mos, Power.Russia),
    new Unit(Army, $.War, Power.Russia), new Unit(Fleet, $.StP_SC, Power.Russia),
    new Unit(Army, $.Smy, Power.Turkey), new Unit(Army, $.Con, Power.Turkey),
    new Unit(Fleet, $.Ank, Power.Turkey)
  ],
  [],
  [...map.provinces].map(p => {
    if (p.homeOf) {
      return [p, new ProvinceStatus(p.homeOf, false)]
    } else {
      return null
    }
  }).filter(x => x)
)

const variant = new Variant(new Rule(), board)

module.exports = variant

File: core/shared/src/main/scala/diplomacy/variant/standard/Variant.scala
package diplomacy.variant.standard

import scala.scalajs.js.annotation.JSExport

import diplomacy.board.{ DiplomacyUnit, Province }
import diplomacy.variant.{ Variant => BaseVariant }
import diplomacy.variant.standard.map.Power
import diplomacy.variant.standard.map.Keywords._
import diplomacy.variant.standard.board.Keywords._
import diplomacy.variant.standard.board.Implicits._
import diplomacy.variant.standard.rule.{ Rule, Board, ProvinceStatus }
import diplomacy.variant.standard.rule.Keywords._
import diplomacy.variant.standard.rule.Implicits._

@JSExport
object Variant {
  private[this] val initialBoard =
    new Board[board.Turn, Power](
      map.map,
      1901.Spring - Movement,
      units = Set(
        DiplomacyUnit(Austria, Army, Vie), DiplomacyUnit(Austria, Army, Bud), DiplomacyUnit(Austria, Fleet, Tri),
        DiplomacyUnit(England, Army, Lvp), DiplomacyUnit(England, Fleet, Edi), DiplomacyUnit(England, Fleet, Lon),
        DiplomacyUnit(France, Army, Mar),  DiplomacyUnit(France, Army, Par),  DiplomacyUnit(France, Fleet, Bre),
        DiplomacyUnit(Germany, Army, Ber), DiplomacyUnit(Germany, Army, Mun), DiplomacyUnit(Germany, Fleet, Kie),
        DiplomacyUnit(Italy, Army, Ven),   DiplomacyUnit(Italy, Army, Rom),   DiplomacyUnit(Italy, Fleet, Nap),
        DiplomacyUnit(Russia, Army, Mos),  DiplomacyUnit(Russia, Army, War),
        DiplomacyUnit(Russia, Fleet, Sev), DiplomacyUnit(Russia, Fleet, StP_SC),
        DiplomacyUnit(Turkey, Army, Smy),  DiplomacyUnit(Turkey, Army, Con),  DiplomacyUnit(Turkey, Fleet, Ank)
      ),
      Map(),
      provinceStatuses = (map.map.provinces collect {
        case p@ Province(_, Some(home), _) => p -> ProvinceStatus[Power](Option(home), false)
      }).toMap
  )

  private[this] val rule = new Rule[board.Turn, map.Power](board.Turn.nextTurn)

  @JSExport
  val variant = BaseVariant(rule, initialBoard)
}
